### PR TITLE
Set up ESLint, Prettier, and VS Code extensions correctly (+lint code)

### PR DESCRIPTION
**Pull request by snario**
1. Installs `eslint-plugin-prettier` to ensure `eslint` is all we need for formatting too
2. Updates `eslintrc.js` as recommended by the plugin

This makes auto-save also handle formatting using the `ESLint` formatter / extension.

### CODE REVIEW

**Review by maurelian — APPROVED** [review]

**Comment by maurelian** [comment]
FYI, this didn't work for me at first, but it works after: 

```
rm -rf node_modules
yarn install
```
And then reloading vscode.

**Comment by maurelian** [comment]
Based on all these ["checks with annotations in unchanged files"](https://github.com/ethereum-optimism/optimism/pull/1145/files#diff-d83cbd6825ff9cb509323474d30f9df3ec1882e27b902583d48f5c1b81f85bcb), it looks like our `yarn lint:fix` and `yarn lint:check` commands are out of sync. These are the actual commands running in the contracts package: 

```json
    "lint:fix:typescript": "prettier --config .prettierrc.json --write \"hardhat.config.ts\" \"{src,test}/**/*.ts\"",
    "lint:typescript": "eslint -c .eslintrc.js --ext .ts --format stylish .",    
```

**Comment by maurelian** [comment]
But they're just warnings, I say we merge it.

**Comment by snario** [comment]
I installed all the relevant packages to the monorepo root.

**Review by annieke — DISMISSED** [review]
tested a couple typescript packages and this correctly shows warnings and formats on save

File: .vscode/extensions.json
Comment (smartcontracts) [inline .vscode/extensions.json:5]: Are comments allowed in json?

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts
Comment (smartcontracts) [inline packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:393]: Hmmm not sure if I *love* this but I also don't particularly care either way.

**Review by smartcontracts — APPROVED** [review]
Looks good, I'll give an approval but I would recommend waiting for someone else to take a look at the format change introduced in the batch submitter (cc @karlfloersch maybe?) to see if they like it or not.

File: .vscode/extensions.json
Comment (snario) [reply .vscode/extensions.json:5]: In this file specifically, VS Code is okay with it

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts
Comment (snario) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:393]: We can change the rule if you'd like, but it should out of scope of this PR. The reason this diff was in the PR was because this file was saved and our current formatting rules (i.e., prettier) dictate this style and this PR turns on "formatOnSave"

File: .vscode/extensions.json
Comment (smartcontracts) [reply .vscode/extensions.json:5]: Weird, ok!

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts
Comment (smartcontracts) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:393]: I see. Do you mind making another PR after this is merged that runs `prettier`?

Comment (snario) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:393]: Yes. We'll need to co-ordinate though as many files will be changed. We may want to try to modify the prettier config in such a way to minimize this formatting diff.

**Comment by snario** [comment]
Linted everything in 4027fdf

File: integration-tests/package.json
Comment (elenadimitrova) [inline integration-tests/package.json:38]: Are there are obsoleted references to `prettier` we could remove?

File: .eslintrc.js
Comment (elenadimitrova) [inline .eslintrc.js:7]: This exclusion for `hardhat` is no longer working :/

File: packages/batch-submitter/package.json
Comment (elenadimitrova) [inline packages/batch-submitter/package.json:16]: This pre-commit hook here isn't producing the same results as running `lint:check`. I am testing this by messing with the `optimism/packages/contracts/test/contracts/OVM/bridge/base/OVM_L1CrossDomainMessenger.spec.ts` test file.

Comment (snario) [reply packages/batch-submitter/package.json:16]: To test this I made the following diff:

```diff
i/packages/contracts/test/contracts/OVM/bridge/base/OVM_L1CrossDomainMessenger.spec.ts
+++ w/packages/contracts/test/contracts/OVM/bridge/base/OVM_L1CrossDomainMessenger.spec.ts
@@ -23,11 +23,10 @@ import {
 import { keccak256 } from 'ethers/lib/utils'
 import { predeploys } from '../../../../../src'
 
-const MAX_GAS_LIMIT = 8_000_000
+let MAX_GAS_LIMIT = 8_000_000
 
 const deployProxyXDomainMessenger = async (
-  addressManager: Contract,
-  l1XDomainMessenger: Contract
+  addressManager: Contract,l1XDomainMessenger: Contract
 ): Promise<Contract> => {
   await addressManager.setAddress(
     'OVM_L1CrossDomainMessenger',
```

Then ran `yarn pre-commit` inside the `contracts` package. The output is:

```
/Users/liam/Projects/optimism/optimism/packages/contracts/test/contracts/OVM/bridge/base/OVM_L1CrossDomainMessenger.spec.ts
  26:5   error    'MAX_GAS_LIMIT' is never reassigned. Use 'const' instead  prefer-const
  29:28  warning  Insert `⏎··`   
```

Then I ran `yarn lint:check` and I see

```
/Users/liam/Projects/optimism/optimism/packages/contracts/test/contracts/OVM/bridge/base/OVM_L1CrossDomainMessenger.spec.ts
  26:5   error    'MAX_GAS_LIMIT' is never reassigned. Use 'const' instead  prefer-const
  29:28  warning  Insert `⏎··`                                              prettier/prettier
```

Which seems to match — do you have an example of something not matching?

File: .eslintrc.js
Comment (snario) [reply .eslintrc.js:7]: How are determining this?

File: integration-tests/package.json
Comment (snario) [reply integration-tests/package.json:38]: I don't think so — we still use prettier it's just used within `eslint` as a plugin now.

File: integration-tests/package.json
Comment (snario) [reply integration-tests/package.json:11]: I think in `specs` it isn't unreasonable to keep using `prettier` since it's being used for Markdown linting as opposed to JS or TS and it would likely be overkill to use ESLint for that purpose

File: integration-tests/package.json
Comment (snario) [reply integration-tests/package.json:38]: I don't think so — we still use `prettier` inside `eslint` as a plugin

**Review by smartcontracts — APPROVED** [review]
Approving because I can't see anything obviously incorrect. Can re-approve if more commits are added, just LMK.

File: packages/batch-submitter/package.json
Comment (elenadimitrova) [reply packages/batch-submitter/package.json:16]: This is already merged before I had a chance to test again but we'll find out soon I guess

**Pull request merged**
